### ISSUE TITLE
error code too long for 32bit

### ISSUE THREAD

**Issue by wstidolph** [ga4gh/server#443]
The common.avdl shows a GAException errorCode is an int, which in Avro 1.7.7 (as in Java) is a signed 32 bit value(maxint value is 2^31 - 1, or  2147483647) . But I'm getting back error codes from the server that don't fit:
```
curl --data '{"readGroupIds":["low-coverage:HG00533 --header 'Content-Type: application/json' http://localhost:8000/v0.5.1/reads/search
{"errorCode": 2852766030, "message": "readGroupId 'low-coverage:HG00533' not found"}
```
then when I try to parse that block I get:
```
com.google.gson.JsonSyntaxException: java.lang.NumberFormatException: Expected an int but was 2852766030 at line 1 column 24 path $.errorCode
```
(Using the google gson lib to parse the avro-JSON)

**Comment by dcolligan** [comment]
Here's the code that computes exception codes in `exceptions.py`:

```
    @classmethod
    def getErrorCode(cls):
        """
        Returns an integer that can be used to identify this class.
        """
        # We use the CRC32 digest of the class name as a unique code.
        # We follow the recommendation of the Python docs to ensure
        # that this value is signed 32 bit integer.
        code = zlib.crc32(cls.__name__) & 0xffffffff
        return code
```

It seems like the `zlib` call returns an int in the range `[0, 2**32-1]`.  One solution to this issue would be to subtract `2**31` from what is now currently computed, and return a value that may be negative.  Returning negative error codes does seem a bit weird though...

https://docs.python.org/2/library/zlib.html#zlib.crc32

**Comment by dcolligan** [comment]
Another option would be to `%` the currently computed number by `2**31-1`, but this halves the possible values that could be returned.  Which may not be the biggest problem in the world...

**Comment by jeromekelleher** [comment]
Can we just divide the number by 2 @dcolligan? I don't think we'll run out of codes, and the tests will tell us if we have any collisions in the defined exceptions.

**Comment by wstidolph** [comment]
The error code is just the hash of the name of the class that generates the error? Doesn't that mean we can't track back from the code to anything (or do we keep a reverse-lookup dictionary somewhere?), that renaming a class changes the error codes, and that multiple errors from the same class but with different causes would generate the same code? What's the use for the errorCode?

**Comment by dcolligan** [comment]
We have a test to ensure there are no errorCode collisions.

As of right now, we don't have a worked out plan about how the errorCodes will be used / documented / generated / etc. so this seemed like a good stand in until we do.  We also don't have any backward compatibility requirement, since no one is really using the codes yet.

**Comment by wstidolph** [comment]
I was planning to add "server code is as expected" as a test, but I guess
we don't need it yet :) FWIW, since they're not used, I would put in a "-1"
so it's clear there's no info there.

On Tue, Jun 9, 2015 at 9:00 AM, Danny Colligan <notifications@github.com>
wrote:

> We have a test to ensure there are no errorCode collisions.
>
> As of right now, we don't have a worked out plan about how the errorCodes
> will be used / documented / generated / etc. so this seemed like a good
> stand in until we do. We also don't have any backward compatibility
> requirement, since no one is really using the codes yet.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ga4gh/server/issues/443#issuecomment-110415348>.
>



-- 
  Wayne Stidolph  | (831) 419 7146 | LinkedIn
<http://www.linkedin.com/in/waynestidolph> | Google+
<https://google.com/+WayneStidolph> | twitter
<https://twitter.com/#!/wstidolph> | personal blog
<http://waynedge.blogspot.com/>

**Comment by jeromekelleher** [comment]
@wstidolph my understanding of error codes is that they are implementation specific values used to help diagnose problems. There is no specification of codes between implementations, and no particular requirement that they be used. For us, the code allows us to find out the name of the exception that caused the error, as we have a script to translate the code into an error class name.

**Comment by wstidolph** [comment]
But it's info you're sending to the client so they (me!) expect it to have
some meaning. And it still has the problems of only tracking to a
generating-class name.

And I can't validate its value in a test case because an implementation
refactor can change the classname /hash.

Since you have a code-space, maybe generate an error instance UUID? Every
error occurrence gets/logs unique ID. Use that to search server logs for
trace back info. No big deal if it rolls over every 2^32 errors :)

**Comment by jeromekelleher** [comment]
I don't think you can ever validate this @wstidolph: it's purely an implementation specific value. Each server will have their own way of using the error code, and many probably won't bother setting the codes at all. I agree this is something that should be better specified, but it would need to be done at the schema level. Given the current nebulous description in the schemas, this (mapping to a class name) seemed like the simplest way of making some use of this field.

**Comment by wstidolph** [comment]
OK, no validation use; all I check is that it parses against GA Exception
as a signed int.

**Comment by dcolligan** [comment]
Fixed in #444
